src/ripple/app/tx/impl/NFTokenBurn.cpp

```diff
@@ -104,9 +109,35 @@ NFTokenBurn::doApply()
         view().update(issuer);
     }
 
-    // Optimized deletion of all offers.
-    nft::removeAllTokenOffers(view(), keylet::nft_sells(ctx_.tx[sfNFTokenID]));
-    nft::removeAllTokenOffers(view(), keylet::nft_buys(ctx_.tx[sfNFTokenID]));
+    if (ctx_.view().rules().enabled(fixUnburnableNFToken))
+    {
+        // Delete up to 500 offers, but prefers buy offers first
```

Comment:
team would like to discuss with @ledhed2222

Reply:
test

Reply:
I'm not sure what how severe it is to having the extra directory on the ledger. Down the line, we are looking to implement (from Scott D.)
 
 _When anyone deletes one of their orphaned offers, we should delete as many orphaned offers as we can in that transaction (like as do with expired offers in the payment engine)._ 

So the sell directory is unlikely to hang on the ledger forever, as soon as someone submits a transaction to delete one of their orphaned offers, all the orphaned offers of the burnt deleted will be removed. So I don't think it would be a big problem then.

Reply:
i'm curious - if i have an NFT with 600 buy offers, which I burn, there are now 100 orphaned offers. what parties are able to cancel the orphaned offers at that point, other than the person who owns the offer (and if they are expired, anyone)? presumably I should still be able to cancel them since I am listed as the "owner" on the offer.

but generally, I don't think i have a huge POV/opinion on which offers get cancelled with higher priority. what scott is saying makes sense. i might go with that.

however, one thing i'm wondering is whether we can included a `warning` if the token burn did not successfully cancel all offers. something like `Not all offers for this NFT could be cancelled. You should immediately cancel any remaining offers`.

Reply:
Currently there are two parties that have permission to cancel NFTokenOffers, whether those offers are orphaned or not:

- The `Owner` of the offer can always cancel the offer.
-  _If_ there's a `Destination`, then the `Destination` can cancel the offer.

That's assuming that the offer has not expired, in which case the offer can be canceled by anybody.

Here's the code: https://github.com/XRPLF/rippled/blob/develop/src/ripple/app/tx/impl/NFTokenCancelOffer.cpp#L77-L87

Regarding a `warning`, that's not really something that the ledger internals can provide.  There are three types of responses:

- `tesSUCCESS` is what you get if the transaction succeeded.  If the `NFTokenBurn` succeeds, whether it leaves orphan offers or not, the appropriate response is `tesSUCCESS`.
- `tec*` is the range of responses where the ledger did not do what you asked, but burned your fee anyway.  That's not what is happening here, since the burn was successful.
- The are other error types (`ter`, `ter`, and so forth) to indicate other kinds of failures.  But they all mean that your transaction did not change the ledger in any way.

At any rate, such a warning would be returned to the wrong person.

The person who is leaving the orphaned offers is the issuer burning the token.  They really don't care whether there are left over offers in the ledger.  The only folks who care are the offer owners.  And they only care once they want to recover the reserve on their offers.  So they may not be motivated to cancel their offers for a long time.

Reply:
@seelabs' suggestion may help, but I have doubts about its effectiveness in a larger scale.  Here's why.

Offers on the DEX are encountered whenever someone wants to trade a currency pair.  There are some currency pairs that are very popular, so their order books get a lot of traffic.  There are other currency pairs that see a lot less traffic.  For those low-traffic currencies, dead offers can sit around for a long time before the DEX stumbles across them and removes them.

Compared to currency pairs, there are _lots_ of NFTs on the ledger.  And there are up to two order books for _every individual_ NFT in the ledger.  I would be astonished if _any_ NFTs are traded frequently.  The bulk of NFTs, and offers for them, will just sit on the ledger.  Any single NFToken will probably be traded relatively infrequently.

So I doubt that suggestion will help much.

But at a more meta level, what I hear you saying is that the patch you are providing for this problem is insufficient.  We need yet another patch to clean up what is left behind by this patch.  Amendments are not free.  We should get everything that we think is needed into one amendment.

Reply:
@scottschurr (cc. @seelabs )
>But at a more meta level, what I hear you saying is that the patch you are providing for this problem is insufficient. We need yet another patch to clean up what is left behind by this patch. Amendments are not free. We should get everything that we think is needed into one amendment.

Our team's idea is try to get this change into the next release(1.10), so that we can get voting and hear community's voice as early as possible. The voting may or may not be in our favor, but we want to get it into next release to get a feel what the community likes, and perhaps making more adjustments from there. 

While I do agree that it feels more robust to include the patch of cleaning orphaned offers into the existing one, but unfortunately this what not considered for scoping, and including it would push back the timeline even further.